File: cpplox/source/scanner.cpp
#include <string_view>
#include <unordered_map>

#include "lox/scanner.h"

namespace
{

constexpr bool isDigit(char c)
{
  return c >= '0' && c <= '9';
}

constexpr bool isAlpha(char c)
{
  return (c >= 'a' && c <= 'z') || (c >= 'A' && c <= 'Z') || c == '_';
}
}  // namespace

Scanner::Scanner(std::string_view source)
    : _source {source}
{
  start = _source.begin();
  current = _source.begin();
  line = 1;
}

Scanner::Scanner(const Scanner& other)
    : start {other.start}
    , current(other.current)
    , line {other.line}
{
Scanner& Scanner::operator=(const Scanner& other)
{
  if (this != &other) {
    line = other.line;
    _source = other._source;
    const auto distanceFromBeginning =
        std::distance(other._source.cbegin(), other.start);
    const auto distanceFromStart = std::distance(other.start, other.current);

    start = _source.begin();
    std::advance(start, distanceFromBeginning);

    current = start;
    std::advance(current, distanceFromStart);
  }

  return *this;
}

bool Scanner::isAtEnd() const
{
  return *current == '\0';
}

char Scanner::advance()
{
  current++;
  return current[-1];
}

bool Scanner::match(char expected)
{
  if (isAtEnd()) {
    return false;
  }

  if (*current != expected) {
    return false;
  }

  current++;
  return true;
}

char Scanner::peek() const
{
  return *current;
}

char Scanner::peekNext() const
{
  if (isAtEnd()) {
    return '\0';
  }
  return current[1];
}

void Scanner::skipWhitespace()
{
  while (true) {
    char c = peek();
    switch (c) {
      case ' ':
      case '\r':
      case '\t':
        advance();
        break;
      case '\n':
        line++;
        advance();
        break;
      case '/':
        if (peekNext() == '/') {
          // comments go to the end of the line
          while (peek() != '\n' && !isAtEnd()) {
            advance();
          }
        } else {
          return;
        }
        break;
      default:
        return;
    }
  }
}

Token Scanner::makeToken(TokenType type) const
{
  return Token {
      type, {start, static_cast<unsigned long>(current - start)}, line};
}

Token Scanner::errorToken(std::string_view message) const
{
  return Token {TokenType::ERROR, message, line};
}

TokenType Scanner::identifierType()
{
  std::string_view str {start, static_cast<unsigned long>(current - start)};

  const std::unordered_map<std::string_view, TokenType> keywords {
      {"and", TokenType::AND},
      {"class", TokenType::CLASS},
      {"else", TokenType::ELSE},
      {"false", TokenType::FALSE},
      {"for", TokenType::FOR},
      {"fun", TokenType::FUN},
      {"if", TokenType::IF},
      {"nil", TokenType::NIL},
      {"or", TokenType::OR},
      {"print", TokenType::PRINT},
      {"return", TokenType::RETURN},
      {"super", TokenType::SUPER},
      {"this", TokenType::THIS},
      {"true", TokenType::TRUE},
      {"var", TokenType::VAR},
      {"while", TokenType::WHILE},
  };

  if (keywords.count(str) > 0) {
    return keywords.at(str);
  }

  return TokenType::IDENTIFIER;
}

Token Scanner::string_()
{
  while (peek() != '"' && !isAtEnd()) {
    if (peek() == '\n') {
      line++;
    }
    advance();
  }

  if (isAtEnd()) {
    return errorToken("Unterminated string.");
  }

  // closing quote
  advance();
  return makeToken(TokenType::STRING);
}

Token Scanner::number()
{
  while (isDigit(peek())) {
    advance();
  }

  // Look for fractional part
  if (peek() == '.' && isDigit(peekNext())) {
    // consume the dot
    advance();

    while (isDigit(peek())) {
      advance();
    }
  }

  return makeToken(TokenType::NUMBER);
}

Token Scanner::identifier()
{
  while (isAlpha(peek()) || isDigit(peek())) {
    advance();
  }

  return makeToken(identifierType());
}

Token Scanner::scanToken()
{
  skipWhitespace();
  start = current;

  if (isAtEnd()) {
    return makeToken(TokenType::END_OF_FILE);
  }

  const char c = advance();

  if (isAlpha(c)) {
    return identifier();
  }

  if (isDigit(c)) {
    return number();
  }

  switch (c) {
      // One character tokens
    case '(':
      return makeToken(TokenType::LEFT_PAREN);
    case ')':
      return makeToken(TokenType::RIGHT_PAREN);
    case '{':
      return makeToken(TokenType::LEFT_BRACE);
    case '}':
      return makeToken(TokenType::RIGHT_BRACE);
    case ';':
      return makeToken(TokenType::SEMICOLON);
    case ',':
      return makeToken(TokenType::COMMA);
    case '.':
      return makeToken(TokenType::DOT);
    case '-':
      return makeToken(TokenType::MINUS);
    case '+':
      return makeToken(TokenType::PLUS);
    case '/':
      return makeToken(TokenType::SLASH);
    case '*':
      return makeToken(TokenType::STAR);

      // Two character tokens
    case '!':
      return makeToken(match('=') ? TokenType::BANG_EQUAL : TokenType::BANG);
    case '=':
      return makeToken(match('=') ? TokenType::EQUAL_EQUAL : TokenType::EQUAL);
    case '<':
      return makeToken(match('=') ? TokenType::LESS_EQUAL : TokenType::LESS);
    case '>':
      return makeToken(match('=') ? TokenType::GREATER_EQUAL
                                  : TokenType::GREATER);

    // string literals
    case '"':
      return string_();
  }

  return errorToken("Unexpected character.");
}

File: loxide/src/scanner.rs
use std::rc::Rc;

use crate::token::{Token, TokenType};

pub struct Scanner {
    source: Rc<String>,
    line: usize,
    start: usize,
    current: usize,
}

impl Scanner {
    pub fn new(source: Rc<String>) -> Self {
        return Scanner {
            source,
            line: 1,
            start: 0,
            current: 0,
        };
    }

    pub fn scan_token(&mut self) -> Token {
        self.skip_whitespace();

        self.start = self.current;

        if self.is_at_end() {
            return self.make_token(TokenType::EOF);
        }

        let c = self.advance();

        if c.is_ascii_digit() {
            return self.number();
        }

        if is_alphanumeric(c) {
            return self.identifier();
        }

        match c {
            '(' => {
                return self.make_token(TokenType::LEFT_PAREN);
            }
            ')' => {
                return self.make_token(TokenType::RIGHT_PAREN);
            }
            '{' => {
                return self.make_token(TokenType::LEFT_BRACE);
            }
            '}' => {
                return self.make_token(TokenType::RIGHT_BRACE);
            }
            ';' => {
                return self.make_token(TokenType::SEMICOLON);
            }
            ',' => {
                return self.make_token(TokenType::COMMA);
            }
            '.' => {
                return self.make_token(TokenType::DOT);
            }
            '-' => {
                return self.make_token(TokenType::MINUS);
            }
            '+' => {
                return self.make_token(TokenType::PLUS);
            }
            '/' => {
                return self.make_token(TokenType::SLASH);
            }
            '*' => {
                return self.make_token(TokenType::STAR);
            }
            '!' => {
                let t = if self.match_char('=') {
                    TokenType::BANG_EQUAL
                } else {
                    TokenType::BANG
                };
                return self.make_token(t);
            }
            '=' => {
                let t = if self.match_char('=') {
                    TokenType::EQUAL_EQUAL
                } else {
                    TokenType::EQUAL
                };
                return self.make_token(t);
            }
            '<' => {
                let t = if self.match_char('=') {
                    TokenType::LESS_EQUAL
                } else {
                    TokenType::LESS
                };
                return self.make_token(t);
            }
            '>' => {
                let t = if self.match_char('=') {
                    TokenType::GREATER_EQUAL
                } else {
                    TokenType::GREATER
                };
                return self.make_token(t);
            }
            '"' => {
                return self.string();
            }
            _ => {}
        }
        return self.error_token("Unexpected character.");
    }

    fn is_at_end(&self) -> bool {
        return self.current == self.source.len();
    }

    fn error_token(&self, error: &str) -> Token {
        return Token::new(TokenType::ERROR, error.to_string(), self.line);
    }

    fn make_token(&self, ttype: TokenType) -> Token {
        let substr = self
            .source
            .chars() // TODO: make this handle unicode properly
            .skip(self.start)
            .take(self.current - self.start)
            .collect();

        return Token::new(ttype, substr, self.line);
    }

    fn advance(&mut self) -> char {
        self.current += 1;
        return self.source.chars().nth(self.current - 1).unwrap(); // TODO: are iterators fast? is there a way to index strings directly?
    }

    fn match_char(&mut self, expected: char) -> bool {
        if self.is_at_end() {
            return false;
        }

        let nth_char = self.source.chars().nth(self.current);

        match nth_char {
            None => return false,
            Some(c) => {
                if c != expected {
                    return false;
                }
            }
        }

        self.current += 1;
        return true;
    }

    fn skip_whitespace(&mut self) {
        loop {
            let c = self.peek();

            match c {
                ' ' | '\r' | '\t' => {
                    self.advance();
                }
                '\n' => {
                    self.line += 1;
                    self.advance();
                }
                '/' => {
                    if self.peek_next() == '/' {
                        while self.peek() != '\n' && !self.is_at_end() {
                            self.advance();
                        }
                    } else {
                        return;
                    };
                }
                _ => return,
            };
        }
    }

    fn peek(&self) -> char {
        return self.source.chars().nth(self.current).unwrap();
    }

    fn peek_next(&self) -> char {
        match self.is_at_end() {
            true => return '\0',
            false => {
                return self
                    .source
                    .chars()
                    .nth(self.current + 1)
                    .expect("Source terminated early!")
            }
        }
    }

    fn string(&mut self) -> Token {
        while self.peek() != '"' && !self.is_at_end() {
            if self.peek() == '\n' {
                self.line += 1;
            }
            self.advance();
        }

        if self.is_at_end() {
            return self.error_token("Unterminated string.");
        }

        // closing quote
        self.advance();
        return self.make_token(TokenType::STRING);
    }

    fn number(&mut self) -> Token {
        while self.peek().is_ascii_digit() {
            self.advance();
        }

        // fractional part
        if self.peek() == '.' && self.peek_next().is_ascii_digit() {
            self.advance(); // consume the '.'

            while self.peek().is_ascii_digit() {
                self.advance();
            }
        }

        return self.make_token(TokenType::NUMBER);
    }

    fn identifier(&mut self) -> Token {
        while is_alphanumeric(self.peek()) || self.peek().is_ascii_digit() {
            self.advance();
        }

        return self.make_token(self.identifier_type());
    }

    fn identifier_type(&self) -> TokenType {
        match self.source.chars().next().unwrap() {
            'a' => {
                return self.check_keyword(1, 2, "nd", TokenType::AND);
            }
            'c' => {
                return self.check_keyword(1, 4, "lass", TokenType::CLASS);
            }
            'e' => {
                return self.check_keyword(1, 3, "lse", TokenType::ELSE);
            }
            'i' => {
                return self.check_keyword(1, 1, "f", TokenType::IF);
            }
            'n' => {
                return self.check_keyword(1, 2, "il", TokenType::NIL);
            }
            'o' => {
                return self.check_keyword(1, 1, "r", TokenType::OR);
            }
            'p' => {
                return self.check_keyword(1, 4, "rint", TokenType::PRINT);
            }
            'r' => {
                return self.check_keyword(1, 5, "eturn", TokenType::RETURN);
            }
            's' => {
                return self.check_keyword(1, 4, "uper", TokenType::SUPER);
            }
            'v' => {
                return self.check_keyword(1, 2, "ar", TokenType::VAR);
            }
            'w' => {
                return self.check_keyword(1, 4, "hile", TokenType::WHILE);
            }
            'f' => {
                if self.current - self.start > 1 {
                    match self.source.chars().nth(1).unwrap() {
                        'a' => {
                            return self.check_keyword(2, 3, "lse", TokenType::FALSE);
                        }
                        'o' => {
                            return self.check_keyword(2, 1, "r", TokenType::FOR);
                        }
                        'u' => {
                            return self.check_keyword(2, 1, "n", TokenType::FUN);
                        }
                        _ => {}
                    }
                }
            }
            't' => {
                if self.current - self.start > 1 {
                    match self.source.chars().nth(1).unwrap() {
                        'h' => {
                            return self.check_keyword(2, 2, "is", TokenType::THIS);
                        }
                        'r' => {
                            return self.check_keyword(2, 2, "ue", TokenType::TRUE);
                        }
                        _ => {}
                    }
                }
            }
            _ => {}
        }

        return TokenType::IDENTIFIER;
    }

    fn check_keyword(&self, start: usize, len: usize, rest: &str, ttype: TokenType) -> TokenType {
        if self.current - self.start == start + len {
            let source_substr = &self.source[self.start + start..self.start + start + len];
            if rest == source_substr {
                return ttype;
            }
        }

        return TokenType::IDENTIFIER;
    }
}

const fn is_alphanumeric(c: char) -> bool {
    return (c >= 'a' && c <= 'z') || (c >= 'A' && c <= 'Z') || c == '_';
}
